### PR TITLE
Add pre-commit config

### PR DESCRIPTION
**Pull request by ashnair1**
While working on `torchgeo` and other projects, I found using [pre-commit](https://pre-commit.com/) simplifies catching style and formatting issues. Sometimes I forget to run isort/black/flake8/pydocstyle and it's only caught in a manual review or the CI workflow.  pre-commit would prevent devs from doing so. 

Pros:
- Optional dependency on dev end
- Easy catching of style and formatting issues

Cons:
- One more file to maintain. Relevant changes in `pyproject.toml` for isort/black/flake8/pydocstyle will need to be reflected in the pre-commit config as well. Hence why I've pegged the libraries to their minimum supported versions.

In the event, a dev really wants to push a commit through they can simply pass `--no-verify` option while committing. 

This is purely optional by the way. So please feel free to close this PR if it doesn't really add anything to the ecosystem.

### CODE REVIEW

**Comment by calebrob6** [comment]
In before @adamjstewart says to add a newline at the end of the file :)

Seriously, thanks @ashnair1! I think this is a good idea to add (e.g. it would prevent me from failing style tests every other commit). I already hacked something like this together in the Makefile, so if we use this we could get rid of the Makefile.

My only concern is this line `additional_dependencies: [torch>=1.7, pytorch-lightning>=1.3, pytest>=6, omegaconf>=2.1]`. Why do we need to specify dependencies here?

**Comment by isaaccorley** [comment]
+1 for pre-commit

**Comment by ashnair1** [comment]
> In before @adamjstewart says to add a newline at the end of the file :)
> 
> Seriously, thanks @ashnair1! I think this is a good idea to add (e.g. it would prevent me from failing style tests every other commit). I already hacked something like this together in the Makefile, so if we use this we could get rid of the Makefile.
> 
> My only concern is this line `additional_dependencies: [torch>=1.7, pytorch-lightning>=1.3, pytest>=6, omegaconf>=2.1]`. Why do we need to specify dependencies here?

Awesome!

Regarding your question, since each hook is run in an isolated environment, we need to explicitly add any typed dependencies. Refer [here](https://github.com/pre-commit/mirrors-mypy)

**Comment by ashnair1** [comment]
> +1 for pre-commit
> 
> I think the only thing I dislike about this is that mypy will take up 99% of the run time but I can just run flake8 separately to work out the linting issues.

Hi @isaaccorley, mypy is definitely the slowest hook but I didn't notice a drastic loss of speed. Could you elaborate on this concern? Why would running flake8 outside pre-commit be faster?

**Comment by adamjstewart** [comment]
I'm personally opposed to using pre-commit hooks for the reason that @isaaccorley mentioned. Imagine the following workflow:

1. Add a new feature
2. Try to commit it, find out that mypy isn't happy
3. Fix the issue, run mypy to confirm that it is fixed
4. Commit it

Not only does this workflow run mypy 3 times, it also runs black/flake8/isort/pydocstyle 2 times even though we already know that they pass. 

With this change, every single commit I try to make takes at least 15 seconds, and much longer if a mypy cache doesn't already exist. This time will only increase as the library gets larger. Also, I don't see a reason to test these particular things and not pytest and sphinx, which will take significantly longer.

While pre-commit hooks are useful if you want to make it impossible to commit code that isn't formatted correctly, we have CI and extensive tests to prevent a PR from being merged with failing tests, so the only advantage of pre-commit hooks is when I commit a change and push directly to main, which I try to avoid unless it's a very small change.

P.S. If we do add this, we should remove `Makefile` which serves the same purpose.

**Comment by adamjstewart** [comment]
Basically, I think this is useful, but I don't want to use it. If it was opt-in/opt-out that would be better. Can users add this themselves and use it if they want to?

**Comment by ashnair1** [comment]
> Basically, I think this is useful, but I don't want to use it. If it was opt-in/opt-out that would be better. Can users add this themselves and use it if they want to?

The existence of the pre-commit config does not force a user to use it. The checks will only run if you install and enable pre-commit hooks.

**Comment by ashnair1** [comment]
> I'm personally opposed to using pre-commit hooks for the reason that @isaaccorley mentioned. Imagine the following workflow:
> 
> 1. Add a new feature
> 2. Try to commit it, find out that mypy isn't happy
> 3. Fix the issue, run mypy to confirm that it is fixed
> 4. Commit it
> 
> Not only does this workflow run mypy 3 times, it also runs black/flake8/isort/pydocstyle 2 times even though we already know that they pass.
> 
> With this change, every single commit I try to make takes at least 15 seconds, and much longer if a mypy cache doesn't already exist. This time will only increase as the library gets larger. Also, I don't see a reason to test these particular things and not pytest and sphinx, which will take significantly longer.
> 
> While pre-commit hooks are useful if you want to make it impossible to commit code that isn't formatted correctly, we have CI and extensive tests to prevent a PR from being merged with failing tests, so the only advantage of pre-commit hooks is when I commit a change and push directly to main, which I try to avoid unless it's a very small change.
> 
> P.S. If we do add this, we should remove `Makefile` which serves the same purpose.

This is a fair point but consider another example

1. Add a new feature
2. Commit it
3. Wait for the entire CI suite to run which may not run since it requires approval by the maintainers.
4. Get a message saying isort wasn't sorted correctly

Running the entire test suite for each commit is about the same as waiting for mypy to finish its checks. I completely agree that it depends on the dev as you said. But as I mentioned previously, pre-commit is totally optional.

The hooks are only enabled if a user goes through these two steps
- `pip install pre-commit`
- `pre-commit install`

**Comment by adamjstewart** [comment]
Thanks for the clarification! In that case I'm not opposed to adding this. Can you delete the Makefile?

File: .pre-commit-config.yaml
Comment (adamjstewart) [inline .pre-commit-config.yaml:22]: Is this actually needed? I assume it still uses the settings in `pyproject.toml` doesn't it?

File: .pre-commit-config.yaml
Comment (adamjstewart) [inline .pre-commit-config.yaml:3]: Is this a `>=` or `==`?

File: .pre-commit-config.yaml
Comment (adamjstewart) [inline .pre-commit-config.yaml:6]: Do package dependencies not get installed?

File: .pre-commit-config.yaml
Comment (adamjstewart) [inline .pre-commit-config.yaml:30]: How did you determine this list of dependencies? A lot of packages have partial type hints, so they should still be installed. The annoying thing about this is that if a dependency gains type hints we won't know. Torchmetrics is definitely missing from this, not sure what else is missing. Personally I would just add all dependencies.

**Review by calebrob6 — APPROVED** [review]

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:22]: `pre-commit` does not use `pyproject.toml` so it is required. Ref: [here](https://github.com/pre-commit/pre-commit/issues/1165)

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:3]: `==`

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:6]: `pre-commit` runs these hooks in isolated environments so any additional dependencies must be specified so that pre-commit knows to install them before running the hook.

Comment (ashnair1) [reply .pre-commit-config.yaml:6]: For example, 

```
-   repo: https://github.com/pycqa/isort
        rev: 5.8.0
        hooks:
        -    id: isort
             additional_dependencies: ["colorama>=0.4.3"]
```

The above config says install repo from `https://github.com/pycqa/isort` with rev/tag `5.8.0` with the additional dependency of `colorama`

File: .pre-commit-config.yaml
Comment (adamjstewart) [reply .pre-commit-config.yaml:22]: Wow, that developer is... not friendly. And this is an extremely obvious feature to have with no downsides whatsoever. Are you sure we actually want to use pre-commit?

File: .pre-commit-config.yaml
Comment (adamjstewart) [reply .pre-commit-config.yaml:3]: That's not very useful since each version of flake8 adds new checks.

File: .pre-commit-config.yaml
Comment (adamjstewart) [reply .pre-commit-config.yaml:6]: How does it install isort then?

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:3]: True. I just specified 3.8 since it's the minimum supported version (for `flake8`) according to `requirements.txt`.

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:6]: At a high level,
- Create environment
- Git clone specific tag and install

Ref: [here](https://pre-commit.com/#pre-commit-configyaml---repos)

Comment (adamjstewart) [reply .pre-commit-config.yaml:6]: > Git clone specific tag and install

What does "install" mean? If they use `pip install` or `python setup.py install` that would install the dependencies as well.

Comment (ashnair1) [reply .pre-commit-config.yaml:6]: Yes but colorama is an optional dependency. Ref: [here](https://github.com/PyCQA/isort/blob/995e016b9d888fea02bedeb5be8ea7f01b694b58/pyproject.toml#L47)

Comment (calebrob6) [reply .pre-commit-config.yaml:6]: "The hook repository must be installable via pip install . (usually by either setup.py or pyproject.toml). The installed package will provide an executable that will match the entry – usually through console_scripts or scripts in setup.py."

https://pre-commit.com/#python

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:30]: You're going to love (hate) this :)

I added libraries that I knew that used type hints and confirmed by running the default mypy. I stopped adding dependencies once I saw the mypy hook and default mypy gave the same result.

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:3]: I don't have strong opinions about us using it. For me `pre-commit` was a convenient tool to catch style and formatting errors on my side rather than push it to the CI workflow and find out about it later. My original goal was to provide this option should any current/future contributors feel the same way and **if** it's harmless to maintain, add it to the repo.

As I mentioned earlier the one con to `pre-commit` is that any major changes to `pyproject.toml` will need to be reflected in the `pre-commit` config. 

Updated black version? Need to do that here. Added a new dependency that supports type hints? Need to add that here.
To be fair, this can be caught pretty easily but yeah this is something to be considered.

Truth be told, if the final call is to close this PR and refer users to it when new contributors keep hitting style/formatting errors that honestly fine too.

File: .pre-commit-config.yaml
Comment (ashnair1) [reply .pre-commit-config.yaml:22]: I don't have strong opinions about us using it. For me `pre-commit` was a convenient tool to catch style and formatting errors on my side rather than push it to the CI workflow and find out about it later. My original goal was to provide this option should any current/future contributors feel the same way and **if** it's harmless to maintain, add it to the repo.

As I mentioned earlier the one con to `pre-commit` is that any major changes to `pyproject.toml` will need to be reflected in the `pre-commit` config. 

Updated black version? Need to do that here. Added a new dependency that supports type hints? Need to add that here.
To be fair, this can be caught pretty easily but yeah this is something to be considered.

Truth be told, if the final call is to close this PR and refer users to it when new contributors keep hitting style/formatting errors that honestly fine too.

**Comment by calebrob6** [comment]
Just wanted to jump in here, @adamjstewart, can we accept this without digging into the details of how pre-commit works (as it isn't going to bother anyone who doesn't use pre-commit)? I'm assuming that whoever uses pre-commit (e,g. at least @ashnair1) will be able to update this file as needed (for example, if it is saying "all good" when CI is saying "style issues").

File: .pre-commit-config.yaml
Comment (adamjstewart) [reply .pre-commit-config.yaml:30]: I'm guessing the reason that not having torchmetrics in that list didn't raise any red flags is that mypy will ignore missing type hints by default. We might just want to add all dependencies (and have a sixth file we need to keep track of all deps in...)

**Comment by isaaccorley** [comment]
One action item that should result from this PR should be to add a couple more lines of detail to the contributing section of the readme. We should explain that one should make sure their PRs use code formatters (isort, mypy), pass all checks (mypy, pydocstyle, flake8), and pass tests (pytest). This would reduce the barrier of entry for potential contributors and clear up any confusion.

**Comment by adamjstewart** [comment]
> can we accept this without digging into the details of how pre-commit works (as it isn't going to bother anyone who doesn't use pre-commit)? I'm assuming that whoever uses pre-commit (e,g. at least @ashnair1) will be able to update this file as needed (for example, if it is saying "all good" when CI is saying "style issues").

Every file we add to the repo is something we promise to maintain. We have no way of testing this particular file, and it is guaranteed to become rapidly out-of-date as `pyproject.toml` is updated, new dependencies are added, or new versions of dependencies are released. I hope that @ashnair1 will continue to use and develop torchgeo far into the future, but if not I'm unable to maintain this file myself.

My biggest concern is that this PR adds yet another file in which we need to list our dependencies:

* setup.cfg
* requirements.txt
* docs/requirements.txt
* spack.yaml
* environment.yml
* .github/workflows/{docs/release/style/tests}.yaml
* .pre-commit-config.yaml

This is no fault of @ashnair1, the tools we use (setuptools, pip, sphinx, spack, conda, github actions, git, etc.) don't know how to communicate with each other. This is also the reason I didn't add tox support (tox provides similar functionality and has the same issues).

So far I have seen many PRs to add new dependencies, and many PRs to add new files in which we need to keep track of our dependencies, but no PRs to try to reduce the number of either of these. There is hope (we should look at pbr or setuptools-scm to see if they can help reduce the number of files). If anyone is interested in tackling this at some point, please let me know.

Overall, I'm fine with merging this as is, but I want to make sure everyone understands and acknowledges these limitations.

**Review by adamjstewart — APPROVED** [review]

**Review by calebrob6 — APPROVED** [review]

**Pull request merged**
